noir-projects/aztec-nr/aztec/src/oracle/message_discovery.nr

```diff
@@ -62,13 +63,129 @@ pub struct PublicLogWithTxData {
 
 /// Fetches a public log from the node that has the corresponding `tag`. The tag is the first field in the log's content.
 /// Returns `Option::none` if no such log exists. Throws if more than one log with that tag exists.
+/// Note that the log content fields length in the return value is PUBLIC_LOG_SIZE_IN_FIELDS + 1 because the contract
+/// address is prepended to the content.
 // TODO(#11627): handle multiple logs with the same tag.
 // TODO(#10273): improve contract siloing of logs, don't introduce an extra field.
 pub unconstrained fn get_public_log_by_tag_from_contract(
     tag: Field,
     contract_address: AztecAddress,
-) -> Option<PublicLogWithTxData> {
-    get_public_log_by_tag_from_contract_oracle(tag, contract_address)
+) -> Option<LogWithTxData<PUBLIC_LOG_SIZE_IN_FIELDS + 1>> {
+    let log = get_public_log_by_tag_from_contract_oracle(tag, contract_address);
+
+    if log.is_some() {
+        // Public logs have an extra field at the beginning with the contract address, which we use to verify
+        // that we're getting the logs from the expected contract.
+        // TODO(#10273): improve how contract log siloing is handled
+        assert_eq(
+            log.unwrap_unchecked().log_content.get(0),
+            contract_address.to_field(),
+            "Got a public log emitted by a different contract",
+        );
+    }
+
+    log
+}
+
+/// Fetches a private log from the node that has the corresponding `tag`. The tag is the first field in the log's
+/// content. Returns `Option::none` if no such log exists. Throws if more than one log with that tag exists.
+pub unconstrained fn get_private_log_by_tag_from_contract(
+    tag: Field,
+    contract_address: AztecAddress,
+) -> Option<LogWithTxData<PRIVATE_LOG_SIZE_IN_FIELDS>> {
+    // Note that unlike in the public case, the contract address is not prepended to the log content and instead
+    // the tag is siloed with the contract address.
+    let siloed_tag = compute_siloed_private_log_field(contract_address, tag);
+    let log = get_private_log_by_tag_oracle(siloed_tag);
+
+    if log.is_some() {
+        // Check if the first field of the log matches the siloed tag.
+        assert_eq(
+            log.unwrap_unchecked().log_content.get(0),
+            compute_siloed_private_log_field(contract_address, tag),
+            "Got a private log emitted with unexpected tag",
+        );
+    }
+
+    log
+}
+
+global CONTRACT_ADDRESS_LEN: u32 = 1;
+
+/// Attempts to retrieve a log by its tag from both public and private sources. First checks public logs, falling back
+/// to private logs if no public log is found. Returns a log with content max length equal to MAX_LOG_CONTENT_LEN.
+/// Since public logs are prefixed with contract address but private logs are not, we remove the contract address from
+/// the public log.
+pub unconstrained fn get_log_by_tag_from_contract<let MAX_LOG_CONTENT_LEN: u32>(
```

Comment:
This function is quite ugly but the ugliness is nicely encapsulated here. I didn't bother too much with improving it (e.g. having a utility copying the log_content to BoundedVec of MAX_LOG_CONTENT_LEN max lenght) as it sounds to me like this is likely to change.

Reply:
No longer relevant.